pkg/kubelet/volume/cache/actual_state_of_world.go

```diff
@@ -265,7 +269,42 @@ type mountedPod struct {
 
 func (asw *actualStateOfWorld) MarkVolumeAsAttached(
 	volumeSpec *volume.Spec, nodeName string, devicePath string) error {
-	_, err := asw.AddVolume(volumeSpec, devicePath)
+	volumeName, err := asw.AddVolume(volumeSpec, devicePath)
+	if err != nil {
+		return err
+	}
+	if asw.kubeClient == nil {
+		return nil
+	}
+	// Update the node status in the apiserver because this will prevent
+	// the attach-detach controller from mutating the same volume.
+	// For the same reason, if this fails don't retry, because the
+	// attache-detach controller might've taken the lock. It could fail
+	// for different reasons, eg: kubelet status update; returning an error
+	// should cause the caller to retry.
+	if err := asw.addVolumeToNode(volumeName, nodeName); err != nil {
+		// Rollback the AddVolume above, if we can't record it on the node.
+		asw.MarkVolumeAsDetached(volumeName, nodeName)
+		return err
+	}
+	return nil
+}
+
+func (asw *actualStateOfWorld) addVolumeToNode(volumeName api.UniqueVolumeName, nodeName string) error {
+	node, err := asw.kubeClient.Core().Nodes().Get(nodeName)
+	if err != nil {
+		return fmt.Errorf("error getting node %q: %v", nodeName, err)
+	}
+	if node == nil {
+		return fmt.Errorf("No node instance returned for %q", nodeName)
+	}
+	for _, v := range node.Status.VolumesInUse {
+		if v == volumeName {
+			return nil
+		}
+	}
```

Comment:
If volume doesn't already exist in the list of `VolumesInUse`, where is it added?